File: packages/pigeon/platform_tests/alternate_language_test_plugin/android/src/test/java/com/example/alternate_language_test_plugin/MultipleArityTest.java
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.example.alternate_language_test_plugin;

import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

import com.example.alternate_language_test_plugin.MultipleArity.MultipleArityFlutterApi;
import io.flutter.plugin.common.BinaryMessenger;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import org.junit.Test;

public class MultipleArityTest {
  @Test
  public void subtract() {
    BinaryMessenger binaryMessenger = mock(BinaryMessenger.class);
    doAnswer(
            invocation -> {
              ByteBuffer message = invocation.getArgument(1);
              BinaryMessenger.BinaryReply reply = invocation.getArgument(2);
              message.position(0);
              @SuppressWarnings("unchecked")
              ArrayList<Object> args =
                  (ArrayList<Object>) MultipleArityFlutterApi.getCodec().decodeMessage(message);
              Long arg0 = (Long) args.get(0);
              Long arg1 = (Long) args.get(1);

              Long output = arg0 - arg1;

              ArrayList<Object> wrapped = new ArrayList<Object>();
              wrapped.add(0, output);

              ByteBuffer replyData = MultipleArityFlutterApi.getCodec().encodeMessage(wrapped);
              replyData.position(0);
              reply.reply(replyData);
              return null;
            })
        .when(binaryMessenger)
        .send(anyString(), any(), any());

    MultipleArityFlutterApi api = new MultipleArityFlutterApi(binaryMessenger);
    api.subtract(
        30L,
        20L,
        new MultipleArity.Result<Long>() {
          public void success(Long result) {
            assertEquals(10L, (long) result);
          }

          public void error(Throwable error) {
            assertEquals(error, null);
          }
        });
  }
}

File: packages/pigeon/platform_tests/flutter_null_safe_unit_tests/test/multiple_arity_test.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_unit_tests/multiple_arity.gen.dart';
import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';
import 'multiple_arity_test.mocks.dart';

@GenerateMocks(<Type>[BinaryMessenger])
void main() {
  test('multiple arity', () async {
    final BinaryMessenger mockMessenger = MockBinaryMessenger();
    when(mockMessenger.send(
            'dev.flutter.pigeon.MultipleArityHostApi.subtract', any))
        .thenAnswer((Invocation realInvocation) async {
      final Object input = MultipleArityHostApi.codec
          .decodeMessage(realInvocation.positionalArguments[1])!;
      final List<Object?> args = input as List<Object?>;
      final int x = (args[0] as int?)!;
      final int y = (args[1] as int?)!;
      return MultipleArityHostApi.codec.encodeMessage(<Object>[x - y]);
    });

    final MultipleArityHostApi api =
        MultipleArityHostApi(binaryMessenger: mockMessenger);
    final int result = await api.subtract(30, 10);
    expect(result, 20);
  });
}
